src/Raven.Server/Documents/Queries/Graph/IntersectionQueryStep.cs

```diff
@@ -132,6 +137,13 @@ public ValueTask Initialize()
 
         private ValueTask CompleteInitializationAfterLeft()
         {
+            //At this point we know we are not going to yield results we can skip running right hand side
+            if (GetType().GetGenericArguments().FirstOrDefault() != typeof(Union) && _left.IsEmpty())
```

Comment:
Same for the right 
Also, let's avoid the reflection here. Add a property on the struct that will check this.

Reply:
done